contentOnly patterns: mark patterns as contentOnly by adding metadata.patternName to the root block

ramonjd opened the pull request:
> [!TIP]
> This PR is largish, but intentionally so. It handles the migration from template hooks (which was buggy) and ensures the feature works only when the content experiment is activated.

## What?

This PR:

-  refactors https://github.com/WordPress/gutenberg/pull/72988, which used hooks to parse patterns in templates, to extend the template controllers taking into account the template activate experiment. 
- ensures that template activation is taken into account

These changes ensure that patterns returned in registered theme templates and template parts have pattern metadata, so that themes that use patterns automatically opt into the [content only behaviour](https://github.com/WordPress/gutenberg/issues/71517).

In Core this is done in [WP_REST_Registered_Templates_Controller](https://github.com/wordpress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-registered-templates-controller.php#L3)

## Why?

Extending the class is required as parsing and re-serializing blocks disrupts the render chain, in particular it breaks shortcodes in patterns and also render_block_core/pattern hooks.

See:

- https://core.trac.wordpress.org/ticket/64279
- https://wordpress.slack.com/archives/C02QB2JS7/p1763612179462009

Thanks to @dd32 for helping debug.

## How?

When the `active_templates` experiment is enabled, WordPress uses a different REST controller (Gutenberg_REST_Registered_Templates_Controller) to fetch theme templates. 

Registered theme patterns are returned from the `/wp/v2/registered-templates` endpoint.

Tests have been added to cover both scenarios (with the experiment on and off).

## Testing Instructions

Check out this PR and head over to `/wp-admin/admin.php?page=gutenberg-experiments` to enable the `contentOnly: Make patterns contentOnly by default upon insertion` experiment.

<img width="899" height="93" alt="Screenshot 2025-11-05 at 2 25 21 pm" src="https://github.com/user-attachments/assets/ebe7893a-7b8d-4a35-981d-3a3d9c56c5d2" />

I'm using TT4 to test.

Head over to `/wp-admin/site-editor.php?p=%2Ftemplate`

In the browser network console, you can check the response of `/wp/v2/registered-templates` to see if the template content contains pattern metadata (search for `metadata`)

Now edit the [All Archives template](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/templates/archive.html) (or any that contains a pattern).

In the list view, check that the [post 3 col pattern](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/patterns/posts-3-col.php) is in content only mode, that is, that it only displays text and other blocks (not containers and so on)

Now enable the template activation experiment and test again.

Check that patterns still work by running through the test steps in https://github.com/WordPress/gutenberg/pull/73375

Ensure that with the content only experiment OFF you cannot reproduce these test results.

## Screenshots or screencast <!-- if applicable -->


| Before  | After |
| ------------- | ------------- |
| <img width="393" height="369" alt="Screenshot 2025-11-10 at 12 38 42 pm" src="https://github.com/user-attachments/assets/137b54f5-2986-4afb-a6c4-db4ec29d459c" />  | <img width="369" height="496" alt="Screenshot 2025-11-10 at 12 36 33 pm" src="https://github.com/user-attachments/assets/ee765bfa-0324-4194-8bed-19bed20fbd1b" /> |

ramonjd committed:
Refactors https://github.com/WordPress/gutenberg/pull/72988, which used filters to parse patterns in template, to extend the template controllers taking into account the template activate experiment. This is required as parsing and re-serializing blocks disrupts the render chain, in particular it breaks shortcodes in patterns and also render_block_core/pattern hooks, because the wrapping pattern was not there.

ramonjd committed:
Enhance block pattern resolution by introducing an experiment toggle for 'gutenberg-content-only-pattern-insertion'. This change modifies the `prepare_item_for_response` method in both the block patterns and templates controllers to conditionally resolve pattern blocks based on the experiment's status. Additionally, updates to the test cases ensure the experiment is enabled during tests, improving performance and reliability.

ramonjd committed:
changelog

ramonjd commented on lib/compat/wordpress-7.0/template-activate.php:
```diff
@@ -116,7 +116,7 @@ function gutenberg_maintain_templates_routes() {
 	$controller->register_routes();
 
-	$registered_template_controller = new Gutenberg_REST_Static_Templates_Controller();
+	$registered_template_controller = new Gutenberg_REST_Registered_Templates_Controller();
```
Optional, but matches the Core class before it was reverted. 

Happy to revert this change, it's a bit superfluous, but at least the class matches the route.

andrewserong replied:
It looks like this PR currently adds the additional `Gutenberg_REST_Registered_Templates_Controller` rather than renaming.

Should we either a) retain the existing name to reduce the scope of this PR, or b) remove the static templates controller file?

ramonjd replied:
Ooh! Bad rebase. Thanks for flagging

> retain the existing name to reduce the scope of this PR

This sounds like a good idea for now. I'll fix it up

ramonjd commented on lib/compat/wordpress-7.0/class-gutenberg-rest-block-patterns-controller-7-0.php:
```diff
@@ -53,8 +53,12 @@ public function register_routes() {
+			$blocks = gutenberg_resolve_pattern_blocks( $blocks );
+		} else {
+			$blocks = resolve_pattern_blocks( $blocks );
+		}
```
Should have been like this from the start I think as `gutenberg_resolve_pattern_blocks` adds pattern metadata to blocks even when the experiment was off.

ramonjd replied:
Actually withdraw that: now I realize that if we take this away, patterns inserted when the experiment is off won't work when the experiment is on.

It adds block metadata and is mostly harmless - I think I'll remove this and the associated tests.

ramonjd commented on lib/compat/wordpress-7.0/rest-api.php:
```diff
@@ -7,70 +7,28 @@
-				$blocks            = gutenberg_resolve_pattern_blocks( $blocks );
-				$template->content = serialize_blocks( $blocks );
-			}
+if ( ! gutenberg_is_experiment_enabled( 'active_templates' ) ) {
```
The template activation experiment registers its own routes, so we only register the registered templates controller if the experiment is not enabled.

andrewserong replied:
Sorry, just caught this — I think I've found a subtle bug here. In the `active_templates` experiment it handles the case for `wp_template` so the `patternName` injection happens correctly in the `/wp/v2/templates` endpoint.

However for template parts `wp_template_part` isn't handled by the `active_templates` experiment. So I think we might need to include it here?

To test, I ran the following in my browser console with TT4 theme and took a look at the footer template part, since that one contains a pattern.

```
wp.apiFetch( { path: '/wp/v2/template-parts' } );
```

With the active templates experiment _disabled_ the footer contains the `patternName` as expected:

```
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|50","bottom":"var:preset|spacing|50"}}},"layout":{"type":"constrained"},"metadata":{"patternName":"twentytwentyfour/footer","name":"Footer with colophon, 4 columns","description":"A footer section with a colophon and 4 columns.","categories":["footer"]}} -->
```

With the active templates experiment _enabled_ the `patternName` was not present:

```
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|50","bottom":"var:preset|spacing|50"}}},"layout":{"type":"constrained"}} -->
```

Is that legit, or am I overthinking this?

ramonjd replied:
> So I think we might need to include it here?

Nice work! I suspect you're right. I'll test

<img width="450" height="300" alt="image" src="https://github.com/user-attachments/assets/ebef94cd-70e1-473a-b96d-49e42a1581fb" />

ramonjd replied:
What you're saying makes sense to me, but I can't reproduce in TT4 or TT5 (both footers have patterns). Was there anything else you did?

I enabled the following experiments:

<img width="508" height="400" alt="Screenshot 2025-11-27 at 3 47 02 pm" src="https://github.com/user-attachments/assets/fe0e49be-fa4e-46cd-8b6a-d024d7eb9522" />

And here's the REST API response from with template activation enabled and disabled 🤔 

```html
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|50","bottom":"var:preset|spacing|50"}}},"layout":{"type":"constrained"},"metadata":{"patternName":"twentytwentyfour/footer","name":"Footer with colophon, 4 columns","description":"A footer section with a colophon and 4 columns.","categories":["footer"]}} --> ...
```

Template activation instantiates `Gutenberg_REST_Static_Templates_Controller`, which extends `Gutenberg_REST_Templates_Controller_7_0`. The templates controller is used for [both parts and templates](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/post.php#L402). 

🤔

andrewserong replied:
Have you made sure that you haven't already made any changes to either of those template parts? If you've already saved the post content of either of those template parts, then the `patternName` will be baked in. To find the problem, I had to make sure I'd reset the template parts to emulate a site that hasn't customised the header or footer yet.

The thing that gave me pause is that when the `active_templates` experiment is switched on, it only overrides the `rest_controller_class` for `wp_template` in this block: https://github.com/WordPress/gutenberg/blob/9bc2ea5b97641cd3d2fedb6c47ddff17666c6283/lib/compat/wordpress-7.0/template-activate.php#L56, but not `wp_template_part`. So I'm assuming that with that experiment on, the `$args['rest_controller_class']` for `wp_template_part` will be pointing to the core `'WP_REST_Templates_Controller'` controller and not `'Gutenberg_REST_Templates_Controller_7_0'` or `Gutenberg_REST_Static_Templates_Controller` like we might expect 🤔

ramonjd replied:
Yeah, I hadn't opened the template parts at all. I'll clean my env and try again.

ramonjd replied:
Okay I'm with you! My bad. I was running the wrong version of WP underneath the hood.

Let me get a fix up.

andrewserong replied:
Thank you, glad I wasn't imagining things! 😄

ramonjd replied:
It's 99% of the time me 😄

ramonjd replied:
Should be fixed in [0e0c925](https://github.com/WordPress/gutenberg/pull/73477/commits/0e0c9257ed6fac5f2671d005ae5945a41ca6befd)

andrewserong replied:
Nice work, that's looking good and tidy 👍

ramonjd commented on lib/load.php:
```diff
@@ -50,9 +50,12 @@ function gutenberg_is_experiment_enabled( $name ) {
 
+	// WordPress 7.0 compat - templates only (required here so 6.9 controllers that deal with template activation can inherit from it)
+	require __DIR__ . '/compat/wordpress-7.0/class-gutenberg-rest-templates-controller-7-0.php';
+	require __DIR__ . '/compat/wordpress-7.0/class-gutenberg-rest-registered-templates-controller.php';
```
Unorthodox, but required so that template activation can inherit the functionality

ramonjd replied:
Actually, if template activation isn't making 6.9, then it could be moved to `/compat/7.0` anyway, right?

ramonjd commented on phpunit/class-resolve-patterns-in-template-parts-test.php:
```diff
@@ -0,0 +1,355 @@
+<?php
```
Test overkill? Maybe. 

But given the experiment ping pong and the class juggling with template activation, best to be safe.

ramonjd committed:
Refactor block pattern resolution in Gutenberg REST controllers to consistently resolve pattern blocks using `gutenberg_resolve_pattern_blocks`. Removed conditional checks for the 'gutenberg-content-only-pattern-insertion' experiment in the `prepare_item_for_response` methods. Updated related test cases to remove experiment assertions, enhancing test clarity and performance.

ramonjd committed:
Refactor on top of https://github.com/WordPress/gutenberg/pull/73402

ramonjd committed:
Merge branch 'trunk' into update/pattern-resolution-in-templates

andrewserong commented on lib/compat/wordpress-7.0/class-gutenberg-rest-templates-controller-7-0.php:
```diff
@@ -0,0 +1,408 @@
+ * Note: no changes have been made to this class.
+ * This class extension exists only to override the core route,
+ * and to ensure the gutenberg_resolve_pattern_blocks function is used in the prepare_item_for_response method.
+ * See: https://github.com/WordPress/gutenberg/pull/72988
```
Tiny nit: should this be a reference to this PR rather than the earlier one?

```suggestion
 * See: https://github.com/WordPress/gutenberg/pull/73477
```

ramonjd replied:
yes! i always appreciate a copy edit 🙇🏻

ramonjd committed:
Update lib/compat/wordpress-7.0/class-gutenberg-rest-templates-controller-7-0.php

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

andrewserong commented on lib/compat/wordpress-7.0/class-gutenberg-rest-templates-controller-7-0.php:
```diff
@@ -0,0 +1,408 @@
+		 * in the editor, improving performance.
+		 */
+		$blocks        = parse_blocks( $item->content );
+		$blocks        = gutenberg_resolve_pattern_blocks( $blocks );
```
This is the main change, it seems, to make sure that we're calling the Gutenberg function. Is it worth adding a comment here to flag that?

ramonjd replied:
Yes it is worth it. I'll add it!

andrewserong commented on lib/compat/wordpress-7.0/class-gutenberg-rest-templates-controller-7-0.php:
```diff
@@ -0,0 +1,408 @@
+/**
+ * Core class used to access templates via the REST API for WordPress 7.0.
+ *
+ * Note: no changes have been made to this class.
```
Tiny nit: we could remove this line as it does change things slightly. The next sentence explains that on its own, and possibly flags a little better that this isn't _exactly_ a copy and paste. Alternately, we could update this sentence to say "there are no changes in this class that need to be backported to core".

andrewserong commented on phpunit/class-resolve-patterns-in-template-parts-test.php:
```diff
@@ -0,0 +1,344 @@
+		}
+
+		// Clean up.
+		wp_delete_post( $template_part_id, true );
```
Entirely optional, but the clean up could be called before the first assertion, that way if the test fails, it'll still be cleaned up. Otherwise in a failure situation, this will never be reached.

I noticed the same on the following test — it's likely not too much of an issue so feel free to ignore!

andrewserong reviewed (commented):
Looking pretty good so far! I haven't tested manually yet, but I noticed an issue in the rebase where the `class-gutenberg-rest-static-templates-controller.php` file still exists instead of being renamed. I've left a comment about whether we should retain that name for now, or include the renaming in this PR.

I like the inclusion of the tests, as you mention, while they're a little verbose, good to include some coverage given all the back-and-forth!

ramonjd committed:
Refactor template controller for WordPress 7.0 compatibility

- Replaced the `Gutenberg_REST_Registered_Templates_Controller` with `Gutenberg_REST_Static_Templates_Controller` in the load file.
- Deleted the `class-gutenberg-rest-registered-templates-controller.php` file.
- Updated the `class-gutenberg-rest-static-templates-controller.php` to extend the correct base class.
- Made minor adjustments in the `class-gutenberg-rest-templates-controller-7-0.php` for clarity in the code comments.
- Adjusted the template activation process to reflect the new controller structure.

ramonjd committed:
Clarify comments in Gutenberg REST Templates Controller for WordPress 7.0

- Added a note indicating that there are no changes in this class that need to be backported to core.
- Minor formatting adjustment in the code for improved readability.

ramonjd committed:
Enhance REST API for Template Parts in WordPress 7.0

- Added registration for the Registered Templates Parts REST API routes.
- Updated the `gutenberg_modify_wp_template_post_type_args_7_0` function to simplify its parameters.
- Introduced a new function `gutenberg_modify_wp_template_part_post_type_args_7_0` for handling template part registration.
- Updated PHPUnit tests to verify pattern resolution in template parts via the REST API.

andrewserong reviewed (approved):
Thanks for all the back and forth on this one, this looks good to go to me! Smoke tested to ensure that the contentOnly experiment is still working nicely with patterns contained in template parts (i.e. footer in TT4), and included in templates (e.g. archive in TT4). Toggling the experiment on and off is working as expected.

LGTM! 🚀

ramonjd commented:
Thank you for helping get this one across the line @andrewserong 

Hopefully we won't break wordpress.org 😄 (I tested this against the https://github.com/WordPress/wporg-parent-2021/tree/trunk branch and it looks okay)
